app/src/main/java/unipd/se18/ocrcamera/inci/NameMatchIngredientsExtractor.java

```diff
@@ -0,0 +1,96 @@
+package unipd.se18.ocrcamera.inci;
+
+import android.util.Log;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+
+/**
+ * IngredientsExtractor implementation that for each INCI ingredient check if it is contained
+ * inside the text. Non alphanumeric characters are ignored.
+ * @author Francesco Pham
+ */
+public class NameMatchIngredientsExtractor implements IngredientsExtractor {
+
+    //list where all ingredients from inci db are stored
+    private List<Ingredient> listIngredients;
+
+    //Tag for logs
+    private static final String TAG = "IngredientsExtractor";
+
+    /**
+     * Constructor initializes the ingredients extractor
+     * @param listIngredients Total list of ingredients from the INCI DB
+     */
+    public NameMatchIngredientsExtractor(List<Ingredient> listIngredients) {
+        //copying list so that sorting doesn't affect original list
+        this.listIngredients = new ArrayList<>(listIngredients);
+
+        //sort by name length so when we search for ingredients in text we match longer names first
+        Collections.sort(this.listIngredients, new Comparator<Ingredient>() {
+            @Override
+            public int compare(Ingredient o1, Ingredient o2) {
+                return o2.getInciName().length() - o1.getInciName().length();
+            }
+        });
```

Comment:
I dont understand why we sort the ingredients by length. Is that essential or gives better performance?

Reply:
we sort by length because we want to match longer names first. So for example in the string "SODIUM ACRYLATES COPOLYMER" we don't want to just match "SODIUM" (which also is an ingredient by its own) but the entire ingredient.